cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,483 @@
+					createFunc:   test.expectedGodoFirewallCreateResp,
+					getFunc:      test.expectedGodoFirewallGetResp,
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; want != got {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+		})
+	}
+}
+
+func TestFirewallController_createInboundRules(t *testing.T) {
+	inboundRule := godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "31220",
+	}
+	nodePortService := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeNodePort,
+			Ports: []v1.ServicePort{
+				{
+					Protocol: "tcp",
+					Port:     31220,
+					NodePort: 31220,
+				},
+			},
+		},
+	}
+	fakeServiceList := []*v1.Service{nodePortService}
+
+	testcases := []struct {
+		name        string
+		firewall    *godo.Firewall
+		fwCache     firewallCache
+		serviceList []*v1.Service
+	}{
+		{
+			name:        "successfully updates port range",
+			firewall:    nil,
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+		{
+			name:        "return node port inbound rules when we find a service of type nodeport",
+			firewall:    newFakeFirewall(fakeWorkerFWName, inboundRule),
+			fwCache:     newFakeFirewallCache(inboundRule),
+			serviceList: fakeServiceList,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{},
+			)
+			inboundRules := []godo.InboundRule{inboundRule}
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			rules := fc.createInboundRules(test.serviceList)
+			if !cmp.Equal(rules[0].PortRange, inboundRules[0].PortRange) {
+				t.Errorf("got %q, want %q", rules[0].PortRange, inboundRules[0].PortRange)
+			}
+			if !cmp.Equal(rules[0].Protocol, inboundRules[0].Protocol) {
+				t.Errorf("got %q, want %q", rules[0].Protocol, inboundRules[0].Protocol)
+			}
+		})
+	}
+}
+
+func TestFirewallController_run(t *testing.T) {
+	testcases := []struct {
+		name                           string
+		fwCache                        firewallCache
+		expectedGodoFirewallCreateResp func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallGetResp    func(context.Context, string) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallListResp   func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallUpdateResp func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		fwReconcileFreq                time.Duration
+		expectedError                  error
+	}{
+		{
```

Comment:
Since this test has a single test case only and I don't see us expanding on it for now, I'd suggest to inline it into the test's main thread and update the test method name to be something like `TestFirewallController_NoDataRace`.